<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.533">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Chase M Clark">
<meta name="dcterms.date" content="2024-01-02">
<meta name="description" content="First in a series of posts about what mass spectrometry data looks like, using R">

<title>Computational Pharmacognosy - Introduction to Mass Spectrometry Analysis in R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://hypothes.is/embed.js"></script>
<script>
  window.document.addEventListener("DOMContentLoaded", function (_event) {
    document.body.classList.add('hypothesis-enabled');
  });
</script>
<link href="../../site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="../../site_libs/pagedtable-1.1/js/pagedtable.js"></script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Computational Pharmacognosy</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Posts</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../about/about.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-contributors" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Contributors</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-contributors">    
        <li>
    <a class="dropdown-item" href="../../contributors/chase/chase.html">
 <span class="dropdown-text">Chase</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../contributors/joe/joe.html">
 <span class="dropdown-text">Joe</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/Code4NP/code4np.github.io"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-page-left">
      <h1 class="title">Introduction to Mass Spectrometry Analysis in R</h1>
                  <div>
        <div class="description">
          First in a series of posts about what mass spectrometry data looks like, using R
        </div>
      </div>
                </div>
  </div>
    
  
  <div class="quarto-title-meta column-page-left">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Chase M Clark <a href="https://orcid.org/0000-0001-6439-9397" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">January 2, 2024</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="3">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#the-experiment" id="toc-the-experiment" class="nav-link active" data-scroll-target="#the-experiment">The Experiment</a>
  <ul class="collapse">
  <li><a href="#the-instrument" id="toc-the-instrument" class="nav-link" data-scroll-target="#the-instrument">The Instrument</a>
  <ul class="collapse">
  <li><a href="#ionizer" id="toc-ionizer" class="nav-link" data-scroll-target="#ionizer">Ionizer</a></li>
  <li><a href="#analyzer" id="toc-analyzer" class="nav-link" data-scroll-target="#analyzer">Analyzer</a></li>
  <li><a href="#detector" id="toc-detector" class="nav-link" data-scroll-target="#detector">Detector</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#the-data" id="toc-the-data" class="nav-link" data-scroll-target="#the-data">The Data</a>
  <ul class="collapse">
  <li><a href="#raw-data-formats" id="toc-raw-data-formats" class="nav-link" data-scroll-target="#raw-data-formats">Raw data formats</a></li>
  <li><a href="#open-source-data-formats" id="toc-open-source-data-formats" class="nav-link" data-scroll-target="#open-source-data-formats">Open-source data formats</a>
  <ul class="collapse">
  <li><a href="#mzml" id="toc-mzml" class="nav-link" data-scroll-target="#mzml">mzML</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#analyses-with-r" id="toc-analyses-with-r" class="nav-link" data-scroll-target="#analyses-with-r">Analyses with R</a>
  <ul class="collapse">
  <li><a href="#whats-a-spectrum" id="toc-whats-a-spectrum" class="nav-link" data-scroll-target="#whats-a-spectrum">What’s a spectrum?</a>
  <ul class="collapse">
  <li><a href="#setup-r-session" id="toc-setup-r-session" class="nav-link" data-scroll-target="#setup-r-session">Setup R session</a></li>
  <li><a href="#download-example-data" id="toc-download-example-data" class="nav-link" data-scroll-target="#download-example-data">Download example data</a></li>
  <li><a href="#peek-around" id="toc-peek-around" class="nav-link" data-scroll-target="#peek-around">Peek around</a></li>
  <li><a href="#reanalysis-of-strain-b022-ms-scan-2242-2243-c14-acyl-desferrioxamine" id="toc-reanalysis-of-strain-b022-ms-scan-2242-2243-c14-acyl-desferrioxamine" class="nav-link" data-scroll-target="#reanalysis-of-strain-b022-ms-scan-2242-2243-c14-acyl-desferrioxamine">Reanalysis of strain B022, MS scan 2242 &amp; 2243, C14 acyl-desferrioxamine</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block column-page-left" id="quarto-document-content">





<section id="the-experiment" class="level1">
<h1>The Experiment</h1>
<p>Mass spectrometry has become an integral analytical technique in natural product discovery, both in measuring accurate mass for chemical formula determination, to analyzing molecule fragmentation for structure elucidation and library searches.</p>
<p>I won’t go in to too much detail here but there are some important experimental things to consider when approaching the analysis of mass spectrometry data.</p>
<p>For more info in the components below see <a href="https://www.thermofisher.com/us/en/home/industrial/mass-spectrometry/mass-spectrometry-learning-center/mass-spectrometry-technology-overview.html" target="_blank">this</a> or similar.</p>
<section id="the-instrument" class="level2">
<h2 class="anchored" data-anchor-id="the-instrument">The Instrument</h2>
<section id="ionizer" class="level3">
<h3 class="anchored" data-anchor-id="ionizer">Ionizer</h3>
<p>There are a number of ionizers that are in use in modern mass spectreometers, the most common in outr field being electrospray ionization (ESI). The methods of ionization is important to consider because it will effect which molecules are ionized, how they are ionized, whether the molecules will remain largely intact or fragmented, and what types of adducts you might expect.</p>
</section>
<section id="analyzer" class="level3">
<h3 class="anchored" data-anchor-id="analyzer">Analyzer</h3>
<p>When you think of mass spectrometry you think of being able to differentiate molecules of different masses. Anlyzers are what provide the physical separation (or in the case of orbitraps seperation and measurement) of charged molecules within a mass spectreomter. There are a number of analyzers on the market, with the most popular being quadrupoles, ion traps, orbitraps, time-of-flight, and combinations thereof. The type of analyzer is important to consider in the analysis as well, and the following should be thought about when approaching a new analysis.</p>
<ul>
<li>What is the resolving power of the anlyzer(s)?</li>
<li>Often analyzers will have their efficacy rated in FWHM (full width at half maximum) which is a measure of <a href="https://web.archive.org/web/20230716165120/https://fiehnlab.ucdavis.edu/projects/seven-golden-rules/mass-resolution" target="_blank"><em>resolving power</em></a></li>
<li>If you confuse resolving power with mass resolution you aren’t alone, there’s been much controversy over the years as they somewhat related. See <a href="https://web.archive.org/web/20211130002318/https://www.agilent.com/cs/library/technicaloverviews/public/5991-5885EN.pdf" target="_blank">this whitepaper by Agilent</a>. Simply stated, resolving power measures how well you can separate two mass peaks in a spectrum and resolution is a measure of how “wide” your peaks are.</li>
<li>What is the <a href="https://web.archive.org/web/20230716165120/https://fiehnlab.ucdavis.edu/projects/seven-golden-rules/mass-resolution" target="_blank">scan speed</a> of your analyzer(s)?</li>
<li>Some analyzers are veryt fast (ie they can analyze/separate many m/z per second), while others are slower. Some sacrifice sensitivity, accuracy, resolution, etc for higher scan speed. All things to be aware of.</li>
</ul>
</section>
<section id="detector" class="level3">
<h3 class="anchored" data-anchor-id="detector">Detector</h3>
<p>While there are different detectors, this isn’t usually a concern during the analysis stage. However, if you notice things like sensitivity is too high or low it could be good feedback to give to the instrument operator as it could be detector settings (though likely sample concentration or ionization efficiency).</p>
<p>Another thing to note is that some instruments may have more than one detector and they may serve different purposes. For example, some Time-Of-Flight (TOF) instruments have both a “linear” detector and a “reflectron” detector that elongates the flight path allowing higher resolving power but with a lower m/z ceiling than the “linear” detector. </p><div class="quarto-video"><iframe data-external="1" src="https://www.youtube.com/embed/0jeFpXHZ8W0?start=47" width="250" height="175" title="" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen=""></iframe></div><p></p>
</section>
</section>
</section>
<section id="the-data" class="level1">
<h1>The Data</h1>
<section id="raw-data-formats" class="level2">
<h2 class="anchored" data-anchor-id="raw-data-formats">Raw data formats</h2>
<p>Unfortunately different instrument vendors, and even different instruments within the same vendor, have their own unique data storage format. This is for a variety of reasons, the most convincing being that instruments have ever increasing speeds of ever-increasing data size and proprietary formats that allow faster/better read/write allow a competitive advantage.</p>
</section>
<section id="open-source-data-formats" class="level2">
<h2 class="anchored" data-anchor-id="open-source-data-formats">Open-source data formats</h2>
<p>Fortunately there are open standard formats available. You will proabably encounter mzXML and/or its nwer version, mzML; so if you have the option, choose mzML.</p>
<p>Some vendor software allows converting a file in proprietary data format to mzML, othwise your best bet is likely the program <code>msconvert</code> available as part of the <a href="https://proteowizard.sourceforge.io/download.html" target="_blank">ProteoWizard</a> software library. Unfortunately some vendor formats can only be converted on a Windows computer, a limitation of vendors only providing Windows-based DLLs (so complain your vendor reps about this, not the the ProteoWizard team).</p>
<p><code>msconvert</code> can be used from both a GUI or at the <a href="https://proteowizard.sourceforge.io/tools/msconvert.html" target="_blank">command line</a></p>
<p>For an example of how to use the command line see this zip of a directory that contains a batch file that converts a large number of files at once [https://ccms-ucsd.github.io/GNPSDocumentation/fileconversion/#data-conversion-easy]</p>
<p>I haven’t had the chance to try them but supposedly there are now some <a href="https://github.com/ProteoWizard/container" target="_blank">Docker containers</a> that can successfully run msconvert. If you know how badly this was needed then you know how exciting this would be/is.</p>
<p>I will only cover mzML here as mzXML is similar enough and mzML/mzXML are by far the most commonly encountered in our field. Other formats can be seen at <a href="https://www.psidev.info/specifications" target="_blank">https://www.psidev.info/specifications</a>; and MGF at <a href="http://www.matrixscience.com/help/data_file_help.html" target="_blank">http://www.matrixscience.com/help/data_file_help.html</a></p>
<section id="mzml" class="level3">
<h3 class="anchored" data-anchor-id="mzml">mzML</h3>
<p>So let’s peek at what an mzML file is… As is hinted by the earlier version (mzXML), it is an <a href="https://en.wikipedia.org/wiki/mzXML" target="_blank">XML</a> file, which is a highly-structured “markup language”. You can find the current specifications for the standard as well as example files over at <a href="https://www.psidev.info/mzml" target="_blank">https://www.psidev.info/mzml</a>.</p>
<p>Take a second and look at an example mzml on HUPO-PSI’s <a href="https://github.com/HUPO-PSI/mzML/blob/master/examples/tiny.msdata.mzML0.99.10.mzML" target="_blank">GitHub repo</a>.</p>
<p>The main things to notice are that it is highly structured.</p>
<p>For example:</p>
<pre><code>&lt;dataProcessing id="Xcalibur Processing" softwareRef="Xcalibur"&gt;
&lt;processingMethod order="1"&gt;
&lt;cvParam cvLabel="MS" accession="MS:1000033" name="deisotoping" value="false"/&gt;
&lt;cvParam cvLabel="MS" accession="MS:1000034" name="charge deconvolution" value="false"/&gt;
&lt;cvParam cvLabel="MS" accession="MS:1000035" name="peak picking" value="true"/&gt;
&lt;/processingMethod&gt;
&lt;/dataProcessing&gt;</code></pre>
<ul>
<li><code>&lt;dataProcessing id="Xcalibur Processing" softwareRef="Xcalibur"&gt;</code> opens the “dataProcessing” tag and defines the properties of ‘id’ and ‘softwareRef’; the last line <code>&lt;/dataProcessing&gt;</code> closes the “dataProcessing” tag.</li>
<li>Nested inside within the “dataProcessing” tag is the “processingMethod” tag which opens with <code>&lt;processingMethod order="1"&gt;</code>, and closes with <code>&lt;/processingMethod&gt;</code>.</li>
<li>This contains a further nested tag “cvParam” which is different in that it self-closes with a <code>/&gt;</code>.</li>
</ul>
<p>Usually mzML files are indented which allows you to easily discern which tags are nested under which other tags; but there is no requirement that mzML files have indentations.</p>
</section>
</section>
</section>
<section id="analyses-with-r" class="level1">
<h1>Analyses with R</h1>
<section id="whats-a-spectrum" class="level2">
<h2 class="anchored" data-anchor-id="whats-a-spectrum">What’s a spectrum?</h2>
<p>Let’s finally look at some data. We will use some data I acquired previously to make things easier… on me.</p>
<p>It’s from an LC-MS/MS run of a <a href="https://www.ncbi.nlm.nih.gov/nuccore/1172004435" target="_blank">single <em>Micromonospora</em></a> extract, in a <a href="https://doi.org/10.1073/pnas.1801247115" target="_blank">previously published study</a>.</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Experimental info">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Experimental info
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The sample is a <em>Micromonospora</em> extract. The extraction was performed from a bacterial culture growing on solid A1 agar media following the protocol of Bligh,E. G. and Dyer, W. J. (9). Agar cultures were divided into 1 cm<sup>3</sup> pieces and 3 mm glass beads were added. Extraction solvent was added in three steps with vigorous vortexing between steps 1) 1:2 (v/v) CHCl3:MeOH, 2) CHCl3 in 1/3 the added volume of step one, 3) H2O in 1/3 the added volume of step one. From the resulting two-layer liquid partition, the organic layer was retained for further analysis.</p>
<p>The extract was analyzed via LC-MS/MS with a method adapted from that described by <a href="https://doi.org/10.1021/acscentsci.5b00331" target="_blank">Goering et al</a>. Experiments were performed on an Agilent 1200 workstation connected to a Thermo Fisher Scientific Q-Exactive mass spectrometer with an electrospray ionization source. Reversed-phase chromatography was performed by injection of 20 μL of 0.1 mg/mL of extract at a 0.3 mL/min flow rate across a Phenomenex Kinetex C18 RPLC column (150 mm x 2.1 mm i.d., 2 μm particle size). Mobile phase A was water with 0.1% formic acid and mobile phase B was acetonitrile with 0.1% formic acid. Mobile phase B was held at 15% for 1 minute, then adjusted to 95% over 12 minutes, where it was held for 2 minutes, and the system re-equilibrated for 5 minutes. The mass spectrometry parameters were as follows: scan range 200-2000 m/z, resolution 35,000, scan rate ~3.7 per second. Data were gathered in profile and the top 5 most intense peaks in each full spectrum were targeted for fragmentation that employed a collision energy setting of 25 eV for Higher-energy Collisional Dissociation (HCD) and isolation window of 2.0 <em>m/z</em>.</p>
<p>The mzXML file was created with ProteoWizard’s msconvert, using default settings.</p>
</div>
</div>
</div>
<section id="setup-r-session" class="level3">
<h3 class="anchored" data-anchor-id="setup-r-session">Setup R session</h3>
<p>Install and then load <a href="https://bioconductor.org/packages/release/bioc/html/mzR.html" target="_blank">mzR</a>, a Bioconductor package for parsing mass spectrometry data. Vignette <a href="https://bioconductor.org/packages/release/bioc/vignettes/mzR/inst/doc/mzR.html" target="_blank">here</a>. For plotting I’ll use ggplot2.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"mzR"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"mzR"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"ggplot2"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"ggplot2"</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mzR)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="download-example-data" class="level3">
<h3 class="anchored" data-anchor-id="download-example-data">Download example data</h3>
<p>Next I’ll download the example file to a temporary directory (will be deleted upon closing the R session). Warning: This is a 22 MB and 306 MB download.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>temporary_directory <span class="ot">&lt;-</span> <span class="fu">tempdir</span>()</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>peaks_file_path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(temporary_directory, <span class="st">"B022.mzXML"</span> )</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"ftp://massive.ucsd.edu/v01/MSV000081555/peak/B022.mzXML"</span>,</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> peaks_file_path)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>raw_mzml_path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(temporary_directory, <span class="st">"B022.mzML"</span> )</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"ftp://massive.ucsd.edu/v01/MSV000081555/raw/FullSpectra-mzML/B022_GenbankAccession-KY858245.mzML"</span>,</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> raw_mzml_path)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="peek-around" class="level3">
<h3 class="anchored" data-anchor-id="peek-around">Peek around</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>full_spectra_handle <span class="ot">&lt;-</span> mzR<span class="sc">::</span><span class="fu">openMSfile</span>(raw_mzml_path)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">header</span>(full_spectra_handle)[<span class="dv">2243</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["seqNum"],"name":[1],"type":["int"],"align":["right"]},{"label":["acquisitionNum"],"name":[2],"type":["int"],"align":["right"]},{"label":["msLevel"],"name":[3],"type":["int"],"align":["right"]},{"label":["polarity"],"name":[4],"type":["int"],"align":["right"]},{"label":["peaksCount"],"name":[5],"type":["int"],"align":["right"]},{"label":["totIonCurrent"],"name":[6],"type":["dbl"],"align":["right"]},{"label":["retentionTime"],"name":[7],"type":["dbl"],"align":["right"]},{"label":["basePeakMZ"],"name":[8],"type":["dbl"],"align":["right"]},{"label":["basePeakIntensity"],"name":[9],"type":["dbl"],"align":["right"]},{"label":["collisionEnergy"],"name":[10],"type":["dbl"],"align":["right"]},{"label":["ionisationEnergy"],"name":[11],"type":["dbl"],"align":["right"]},{"label":["lowMZ"],"name":[12],"type":["dbl"],"align":["right"]},{"label":["highMZ"],"name":[13],"type":["dbl"],"align":["right"]},{"label":["precursorScanNum"],"name":[14],"type":["int"],"align":["right"]},{"label":["precursorMZ"],"name":[15],"type":["dbl"],"align":["right"]},{"label":["precursorCharge"],"name":[16],"type":["int"],"align":["right"]},{"label":["precursorIntensity"],"name":[17],"type":["dbl"],"align":["right"]},{"label":["mergedScan"],"name":[18],"type":["int"],"align":["right"]},{"label":["mergedResultScanNum"],"name":[19],"type":["int"],"align":["right"]},{"label":["mergedResultStartScanNum"],"name":[20],"type":["int"],"align":["right"]},{"label":["mergedResultEndScanNum"],"name":[21],"type":["int"],"align":["right"]},{"label":["injectionTime"],"name":[22],"type":["dbl"],"align":["right"]},{"label":["filterString"],"name":[23],"type":["chr"],"align":["left"]},{"label":["spectrumId"],"name":[24],"type":["chr"],"align":["left"]},{"label":["centroided"],"name":[25],"type":["lgl"],"align":["right"]},{"label":["ionMobilityDriftTime"],"name":[26],"type":["dbl"],"align":["right"]},{"label":["isolationWindowTargetMZ"],"name":[27],"type":["dbl"],"align":["right"]},{"label":["isolationWindowLowerOffset"],"name":[28],"type":["dbl"],"align":["right"]},{"label":["isolationWindowUpperOffset"],"name":[29],"type":["dbl"],"align":["right"]},{"label":["scanWindowLowerLimit"],"name":[30],"type":["dbl"],"align":["right"]},{"label":["scanWindowUpperLimit"],"name":[31],"type":["dbl"],"align":["right"]}],"data":[{"1":"2243","2":"2243","3":"2","4":"1","5":"974","6":"431321.5","7":"613.6829","8":"201.1234","9":"117984.9","10":"25","11":"0","12":"51.15211","13":"782.8459","14":"2242","15":"743.5622","16":"1","17":"741449.6","18":"NA","19":"NA","20":"NA","21":"NA","22":"250","23":"FTMS + p ESI d Full ms2 743.56@hcd25.00 [51.67-775.00]","24":"controllerType=0 controllerNumber=1 scan=2243","25":"FALSE","26":"NA","27":"743.5622","28":"1","29":"1","30":"51.66667","31":"775","_rn_":"2243"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">header</span>(full_spectra_handle)[<span class="dv">2243</span>, ]<span class="sc">$</span>precursorMZ</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 743.5622</code></pre>
</div>
</div>
</section>
<section id="reanalysis-of-strain-b022-ms-scan-2242-2243-c14-acyl-desferrioxamine" class="level3">
<h3 class="anchored" data-anchor-id="reanalysis-of-strain-b022-ms-scan-2242-2243-c14-acyl-desferrioxamine">Reanalysis of strain B022, MS scan 2242 &amp; 2243, C14 acyl-desferrioxamine</h3>
<p>2242 is the MS<sup>1</sup> scan that contains the precursor for the MS<sup>2</sup> scan “2243” which appeared in <a href="https://www.pnas.org/doi/full/10.1073/pnas.1801247115#supplementary-materials" target="_blank">Figure S6F</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>single_spectrum <span class="ot">&lt;-</span> mzR<span class="sc">::</span><span class="fu">peaks</span>(full_spectra_handle, <span class="at">scans=</span><span class="dv">2242</span>)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>single_spectrum <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(single_spectrum)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(single_spectrum) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"mass"</span>, <span class="st">"intensity"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(single_spectrum)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 7076</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>  single_spectrum,</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">type =</span> <span class="st">"l"</span>,</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Let’s zoom in on the <sup>13</sup>C isotopic envelope of the “743.5622 <em>m/z</em>” precursor.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> <span class="fu">subset</span>(single_spectrum, mass <span class="sc">&gt;</span> <span class="dv">743</span> <span class="sc">&amp;</span> mass <span class="sc">&lt;</span> <span class="dv">750</span>),</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aes</span>(</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> mass, </span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> intensity</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> </span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">color=</span><span class="st">"gray48"</span>) <span class="sc">+</span> </span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="fl">0.75</span>, <span class="at">color=</span><span class="st">"gray0"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>And zoom in further to the M+H ion.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> <span class="fu">subset</span>(single_spectrum, mass <span class="sc">&gt;</span> <span class="dv">743</span> <span class="sc">&amp;</span> mass <span class="sc">&lt;</span> <span class="dv">744</span>),</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aes</span>(</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> mass, </span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> intensity</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> </span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">color=</span><span class="st">"gray48"</span>) <span class="sc">+</span> </span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="fl">0.75</span>, <span class="at">color=</span><span class="st">"gray0"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>An important thing to take note of when doing most types of spectroscopy/spectrometry is the number of measurements across a peak. Here we are getting ~10 data points per ion/peak which is pretty good. The smaller the number of points, the worse your peak shape will be, the worse your accuracy and precision will be. Alternately, too many points can bloat your data size and sometimes make analyses more difficult. This is largely controlled by <a href="https://www.agilent.com/cs/library/posters/public/Agilent_ASMS_2019_WP444_Poster.pdf" target="_blank">dwell time</a> and duty cycles in the MS acquisition settings. Instrument settings/chromatography may have to be optimized to balance sensitivity with getting enough data points.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> <span class="fu">subset</span>(single_spectrum, mass <span class="sc">&gt;</span> <span class="dv">743</span> <span class="sc">&amp;</span> mass <span class="sc">&lt;</span> <span class="dv">744</span>),</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aes</span>(</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> mass, </span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> intensity</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> </span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">color=</span><span class="st">"gray48"</span>) <span class="sc">+</span> </span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="fl">0.75</span>, <span class="at">color=</span><span class="st">"gray0"</span>) <span class="sc">+</span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> <span class="fu">subset</span>(<span class="fu">subset</span>(single_spectrum, mass <span class="sc">&gt;</span> <span class="dv">743</span> <span class="sc">&amp;</span> mass <span class="sc">&lt;</span> <span class="dv">744</span>), intensity <span class="sc">&gt;</span> <span class="dv">100</span>),<span class="fu">aes</span>(</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> mass, </span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> intensity</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>  ),  <span class="at">size =</span> <span class="dv">3</span>, <span class="at">color=</span><span class="st">"red"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Let’s create a extracted ion chromatogram (EIC) for the “743.5622 <em>m/z</em>” precursor.</p>
<p>To do that we need to loop through all the MS<sup>1</sup> spectra</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>full_spectra_header <span class="ot">&lt;-</span> <span class="fu">header</span>(full_spectra_handle)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>ms1_indices <span class="ot">&lt;-</span> full_spectra_header[full_spectra_header<span class="sc">$</span>msLevel <span class="sc">==</span> <span class="dv">1</span>, ]<span class="sc">$</span>seqNum</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>target_mass <span class="ot">&lt;-</span> <span class="fl">743.5646</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>delta <span class="ot">&lt;-</span> <span class="fl">0.01</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>left_window <span class="ot">&lt;-</span> target_mass <span class="sc">-</span> delta</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>right_window <span class="ot">&lt;-</span> target_mass <span class="sc">+</span> delta</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>z <span class="ot">&lt;-</span> <span class="fu">lapply</span>(ms1_indices, </span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(x){</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>    ret_time <span class="ot">&lt;-</span> full_spectra_header[x, ]<span class="sc">$</span>retentionTime</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> mzR<span class="sc">::</span><span class="fu">spectra</span>(full_spectra_handle, x)</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(x)</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(x) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"mass"</span>, <span class="st">"intensity"</span>)</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> x[x<span class="sc">$</span>mass <span class="sc">&gt;</span> left_window <span class="sc">&amp;</span> x<span class="sc">$</span>mass <span class="sc">&lt;</span> right_window,  ]</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">nrow</span>(x) <span class="sc">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>      <span class="fu">return</span>(<span class="fu">data.frame</span>(<span class="fu">list</span>(<span class="at">ret_time=</span>ret_time, <span class="at">intensity=</span><span class="fu">mean</span>(x<span class="sc">$</span>intensity))))</span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>      <span class="fu">return</span>(<span class="fu">data.frame</span>(<span class="fu">list</span>(<span class="at">ret_time=</span>ret_time, <span class="at">intensity=</span><span class="dv">0</span>)))</span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>z2 <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">"rbind"</span>, z)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Corresponds to <a href="https://www.pnas.org/doi/full/10.1073/pnas.1801247115#supplementary-materials" target="_blank">Figure S5E</a>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>title <span class="ot">=</span> <span class="fu">paste</span>(</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Extracted Ion Chromatogram: "</span>,</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  target_mass,</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">" "</span>,</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">expression</span>(<span class="fu">italic</span>(<span class="st">"m/z"</span>)),</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">" +/- "</span>,</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>  delta,</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>   <span class="st">" Da"</span>)</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> z2,</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aes</span>(</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> ret_time <span class="sc">/</span> <span class="dv">60</span>,</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> intensity</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> </span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">color=</span><span class="st">"gray48"</span>) <span class="sc">+</span></span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Retention Time (min)"</span>) <span class="sc">+</span></span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>      <span class="fu">ggtitle</span>(<span class="fu">bquote</span>(<span class="st">"Extracted Ion Chromatogram:"</span><span class="sc">~</span>.(target_mass) <span class="sc">~</span><span class="fu">italic</span>(<span class="st">"m/z"</span>)<span class="sc">~</span><span class="st">"+/-"</span><span class="sc">~</span>.(delta) <span class="sc">~</span><span class="st">"Da"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Let’s highlight where the instrument fragmented parent ions between 743 <em>m/z</em> &amp; 745 <em>m/z</em>?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> z2,</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">aes</span>(</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> ret_time <span class="sc">/</span> <span class="dv">60</span>,</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> intensity</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> </span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">color=</span><span class="st">"gray48"</span>) <span class="sc">+</span> </span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> <span class="fu">subset</span>(full_spectra_header, precursorMZ <span class="sc">&gt;</span> <span class="dv">743</span> <span class="sc">&amp;</span> precursorMZ <span class="sc">&lt;</span> <span class="dv">745</span>),</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">x=</span>retentionTime <span class="sc">/</span> <span class="dv">60</span>, <span class="at">y=</span> <span class="fl">5e5</span>),</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">color=</span><span class="st">"red"</span></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Retention Time (min)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Another consideration for the experimentalist (and good exam question 😛 ) is how you could obtain more scans of this 743.5622 <em>m/z</em> target ion. There’s multiple ways (the most obvious is to run in targeted mode where you only fragment parent molecules within a tight range around 743.5622 <em>m/z</em>), but if you need untargetted mode you can mess with duty cycles, or adjust your chromatography to increase the elution peak width of the target compound; sometimes 5min chromatography isn’t the best chromatography.</p>
<p>And we can see it is indeed an mzXML file.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="fu">readLines</span>(peaks_file_path, <span class="at">n=</span><span class="dv">10</span>), <span class="at">sep =</span> <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;?xml version="1.0" encoding="ISO-8859-1"?&gt;
&lt;mzXML xmlns="http://sashimi.sourceforge.net/schema_revision/mzXML_3.2"
       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
       xsi:schemaLocation="http://sashimi.sourceforge.net/schema_revision/mzXML_3.2 http://sashimi.sourceforge.net/schema_revision/mzXML_3.2/mzXML_idx_3.2.xsd"&gt;
  &lt;msRun scanCount="4399" startTime="PT0.0673789S" endTime="PT1200.11S"&gt;
    &lt;parentFile fileName="file:///C:\Users\chase\Downloads\LCMSNORTHWESTERN\Example\Input_Folder/20170719_mwm1013_metabologenomics_actinolunaC182x100_B022.raw"
                fileType="RAWData"
                fileSha1="b739a75b1c680e889940f7b35fe9ef07ee5bcd62"/&gt;
    &lt;msInstrument msInstrumentID="1"&gt;
      &lt;msManufacturer category="msManufacturer" value="Thermo Scientific"/&gt;</code></pre>
</div>
</div>
<p>I really like mzR. Under the hood is a lot of fast C/C++ code, and it is possible to lazily load the data from mzML files.</p>
<p>Here we will tell mzR to lazily open the mass spec file we just downloaded. We can see it returns a handle to the file, which contains 4399 “scans”. A scan being a mass spectrum.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>msfile_handle <span class="ot">&lt;-</span> mzR<span class="sc">::</span><span class="fu">openMSfile</span>(peaks_file_path)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>msfile_handle</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Mass Spectrometry file handle.
Filename:  B022.mzXML 
Number of scans:  4399 </code></pre>
</div>
</div>
<p>mzR uses <a href="https://adv-r.hadley.nz/s3.html" target="_blank">S3</a> object oriented programming which is difficult if you are only used to R’s usual functional programming style. You don’t have to worry much about it because most of what I’ll show is functional, but if you do care there are a number of object based methods you can use.</p>
<p>We can see how mzR “opened/parsed” the file, here using C++ code from ProteoWizard.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>msfile_handle<span class="sc">@</span>backend</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>C++ object &lt;0x55f35cb5f5d0&gt; of class 'Pwiz' &lt;0x55f35ac20450&gt;</code></pre>
</div>
</div>
<p>These are the methods available to be run on files parsed with the Pwiz object (not all will be applicable to all types of MS data).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>mzR<span class="sc">:::</span>Pwiz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>C++ class 'Pwiz' &lt;0x55f35ac20450&gt;
Constructors:
    Pwiz()
        docstring : Initialises a new Rccp pwiz object.

Fields: No public fields exposed by this class

Methods: 
     void close()  
           docstring : Close the connection to a mass spec file (mzXML, mzData, etc.).
     void copyWriteMSfile(std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt;, std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt;, std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt;, Rcpp::DataFrame_Impl&lt;Rcpp::PreserveStorage&gt;, Rcpp::List, bool, Rcpp::List)  
           docstring : Copy general content from the originalting MS file, add the provided spectrum list and write the data to a new mzML file.
     Rcpp::Matrix&lt;14, Rcpp::PreserveStorage&gt; get3DMap(std::vector&lt;int, std::allocator&lt;int&gt; &gt;, double, double, double)  
           docstring : Reads all scans and returns them as a matrix.
     Rcpp::DataFrame_Impl&lt;Rcpp::PreserveStorage&gt; getAllChromatogramHeaderInfo()  
           docstring : Returns a data.frame with the header for all chromatograms
     Rcpp::DataFrame_Impl&lt;Rcpp::PreserveStorage&gt; getAllScanHeaderInfo()  
           docstring : Reads the header info for all mass spectra.
     Rcpp::DataFrame_Impl&lt;Rcpp::PreserveStorage&gt; getChromatogramHeaderInfo(Rcpp::IntegerVector)  
           docstring : Returns a data.frame with the chromatogram header information
     Rcpp::DataFrame_Impl&lt;Rcpp::PreserveStorage&gt; getChromatogramsInfo(int)  
           docstring : Reads the chromatogram information.
     std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt; getFilename()  
           docstring : Returns the mass spec filename.
     Rcpp::List getInstrumentInfo()  
           docstring : Reads the instrument information from a pwiz object
     int getLastChrom()  const 
           docstring : Returns the index of the last chromatogram.
     int getLastScan()  const 
           docstring : Returns the last scan (not necessarily the number of scans because of missing scans).
     Rcpp::List getPeakList(Rcpp::IntegerVector)  
           docstring : Performs a non-sequential parsing operation on an indexed mzXML file to obtain the peak list for a numbered scan.
     std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt; getRunStartTimeStamp()  
           docstring : Returns the start time stamp of the MS run.
     Rcpp::DataFrame_Impl&lt;Rcpp::PreserveStorage&gt; getScanHeaderInfo(Rcpp::IntegerVector)  
           docstring : Reads the header info for the specified scan(s). Supports also to extract scan header infos for multiple scans.
     void open(std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt;)  
           docstring : Opens a mass spec file (mzXML, mzData, etc.) and creates a pwiz object
     void writeSpectrumList(std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt;, std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt;, Rcpp::DataFrame_Impl&lt;Rcpp::PreserveStorage&gt;, Rcpp::List, bool, Rcpp::List)  
           docstring : Write the spectrum list to an mzML file.</code></pre>
</div>
</div>
<p>One of the most powerful {{mzR}} functions is <code>header()</code> which provides summarizing information about each scan in the dataset. Each scan is numbered sequentially (seqNum/acquisitionNum) and the msLevel (MS == 1; MS<sup>2</sup> == 2, MS<sup>3</sup> == 3, etc.)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>summary_data <span class="ot">&lt;-</span> <span class="fu">header</span>(msfile_handle)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(summary_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["seqNum"],"name":[1],"type":["int"],"align":["right"]},{"label":["acquisitionNum"],"name":[2],"type":["int"],"align":["right"]},{"label":["msLevel"],"name":[3],"type":["int"],"align":["right"]},{"label":["polarity"],"name":[4],"type":["int"],"align":["right"]},{"label":["peaksCount"],"name":[5],"type":["int"],"align":["right"]},{"label":["totIonCurrent"],"name":[6],"type":["dbl"],"align":["right"]},{"label":["retentionTime"],"name":[7],"type":["dbl"],"align":["right"]},{"label":["basePeakMZ"],"name":[8],"type":["dbl"],"align":["right"]},{"label":["basePeakIntensity"],"name":[9],"type":["dbl"],"align":["right"]},{"label":["collisionEnergy"],"name":[10],"type":["dbl"],"align":["right"]},{"label":["ionisationEnergy"],"name":[11],"type":["dbl"],"align":["right"]},{"label":["lowMZ"],"name":[12],"type":["dbl"],"align":["right"]},{"label":["highMZ"],"name":[13],"type":["dbl"],"align":["right"]},{"label":["precursorScanNum"],"name":[14],"type":["int"],"align":["right"]},{"label":["precursorMZ"],"name":[15],"type":["dbl"],"align":["right"]},{"label":["precursorCharge"],"name":[16],"type":["int"],"align":["right"]},{"label":["precursorIntensity"],"name":[17],"type":["dbl"],"align":["right"]},{"label":["mergedScan"],"name":[18],"type":["int"],"align":["right"]},{"label":["mergedResultScanNum"],"name":[19],"type":["int"],"align":["right"]},{"label":["mergedResultStartScanNum"],"name":[20],"type":["int"],"align":["right"]},{"label":["mergedResultEndScanNum"],"name":[21],"type":["int"],"align":["right"]},{"label":["injectionTime"],"name":[22],"type":["dbl"],"align":["right"]},{"label":["filterString"],"name":[23],"type":["chr"],"align":["left"]},{"label":["spectrumId"],"name":[24],"type":["chr"],"align":["left"]},{"label":["centroided"],"name":[25],"type":["lgl"],"align":["right"]},{"label":["ionMobilityDriftTime"],"name":[26],"type":["dbl"],"align":["right"]},{"label":["isolationWindowTargetMZ"],"name":[27],"type":["dbl"],"align":["right"]},{"label":["isolationWindowLowerOffset"],"name":[28],"type":["dbl"],"align":["right"]},{"label":["isolationWindowUpperOffset"],"name":[29],"type":["dbl"],"align":["right"]},{"label":["scanWindowLowerLimit"],"name":[30],"type":["dbl"],"align":["right"]},{"label":["scanWindowUpperLimit"],"name":[31],"type":["dbl"],"align":["right"]}],"data":[{"1":"1","2":"1","3":"1","4":"1","5":"320","6":"7179157","7":"0.0673789","8":"217.1074","9":"266403.1","10":"NA","11":"0","12":"150.0266","13":"1892.051","14":"NA","15":"NA","16":"NA","17":"NA","18":"NA","19":"NA","20":"NA","21":"NA","22":"0","23":"NA","24":"controllerType=0 controllerNumber=1 scan=1","25":"TRUE","26":"NA","27":"NA","28":"NA","29":"NA","30":"NA","31":"NA","_rn_":"1"},{"1":"2","2":"2","3":"1","4":"1","5":"673","6":"6511908","7":"0.4336310","8":"301.1406","9":"245597.2","10":"NA","11":"0","12":"150.0265","13":"1980.235","14":"NA","15":"NA","16":"NA","17":"NA","18":"NA","19":"NA","20":"NA","21":"NA","22":"0","23":"NA","24":"controllerType=0 controllerNumber=1 scan=2","25":"TRUE","26":"NA","27":"NA","28":"NA","29":"NA","30":"NA","31":"NA","_rn_":"2"},{"1":"3","2":"3","3":"1","4":"1","5":"617","6":"7172014","7":"0.7546260","8":"217.1071","9":"208030.5","10":"NA","11":"0","12":"150.0265","13":"1954.778","14":"NA","15":"NA","16":"NA","17":"NA","18":"NA","19":"NA","20":"NA","21":"NA","22":"0","23":"NA","24":"controllerType=0 controllerNumber=1 scan=3","25":"TRUE","26":"NA","27":"NA","28":"NA","29":"NA","30":"NA","31":"NA","_rn_":"3"},{"1":"4","2":"4","3":"1","4":"1","5":"652","6":"8089850","7":"1.0220100","8":"301.1406","9":"245887.3","10":"NA","11":"0","12":"150.0265","13":"1958.846","14":"NA","15":"NA","16":"NA","17":"NA","18":"NA","19":"NA","20":"NA","21":"NA","22":"0","23":"NA","24":"controllerType=0 controllerNumber=1 scan=4","25":"TRUE","26":"NA","27":"NA","28":"NA","29":"NA","30":"NA","31":"NA","_rn_":"4"},{"1":"5","2":"5","3":"1","4":"1","5":"565","6":"7538651","7":"1.2895000","8":"301.1407","9":"266377.8","10":"NA","11":"0","12":"150.0265","13":"1985.422","14":"NA","15":"NA","16":"NA","17":"NA","18":"NA","19":"NA","20":"NA","21":"NA","22":"0","23":"NA","24":"controllerType=0 controllerNumber=1 scan=5","25":"TRUE","26":"NA","27":"NA","28":"NA","29":"NA","30":"NA","31":"NA","_rn_":"5"},{"1":"6","2":"6","3":"1","4":"1","5":"748","6":"10536195","7":"1.5568800","8":"301.1405","9":"344449.1","10":"NA","11":"0","12":"150.0265","13":"1973.114","14":"NA","15":"NA","16":"NA","17":"NA","18":"NA","19":"NA","20":"NA","21":"NA","22":"0","23":"NA","24":"controllerType=0 controllerNumber=1 scan=6","25":"TRUE","26":"NA","27":"NA","28":"NA","29":"NA","30":"NA","31":"NA","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>So, now we can do things like filtering for only positive mode MS<sup>2</sup> scans.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>filtered_df <span class="ot">&lt;-</span> summary_data[summary_data<span class="sc">$</span>polarity <span class="sc">==</span> <span class="dv">1</span>, ][summary_data<span class="sc">$</span>msLevel <span class="sc">==</span> <span class="dv">2</span>, ]</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(filtered_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["seqNum"],"name":[1],"type":["int"],"align":["right"]},{"label":["acquisitionNum"],"name":[2],"type":["int"],"align":["right"]},{"label":["msLevel"],"name":[3],"type":["int"],"align":["right"]},{"label":["polarity"],"name":[4],"type":["int"],"align":["right"]},{"label":["peaksCount"],"name":[5],"type":["int"],"align":["right"]},{"label":["totIonCurrent"],"name":[6],"type":["dbl"],"align":["right"]},{"label":["retentionTime"],"name":[7],"type":["dbl"],"align":["right"]},{"label":["basePeakMZ"],"name":[8],"type":["dbl"],"align":["right"]},{"label":["basePeakIntensity"],"name":[9],"type":["dbl"],"align":["right"]},{"label":["collisionEnergy"],"name":[10],"type":["dbl"],"align":["right"]},{"label":["ionisationEnergy"],"name":[11],"type":["dbl"],"align":["right"]},{"label":["lowMZ"],"name":[12],"type":["dbl"],"align":["right"]},{"label":["highMZ"],"name":[13],"type":["dbl"],"align":["right"]},{"label":["precursorScanNum"],"name":[14],"type":["int"],"align":["right"]},{"label":["precursorMZ"],"name":[15],"type":["dbl"],"align":["right"]},{"label":["precursorCharge"],"name":[16],"type":["int"],"align":["right"]},{"label":["precursorIntensity"],"name":[17],"type":["dbl"],"align":["right"]},{"label":["mergedScan"],"name":[18],"type":["int"],"align":["right"]},{"label":["mergedResultScanNum"],"name":[19],"type":["int"],"align":["right"]},{"label":["mergedResultStartScanNum"],"name":[20],"type":["int"],"align":["right"]},{"label":["mergedResultEndScanNum"],"name":[21],"type":["int"],"align":["right"]},{"label":["injectionTime"],"name":[22],"type":["dbl"],"align":["right"]},{"label":["filterString"],"name":[23],"type":["chr"],"align":["left"]},{"label":["spectrumId"],"name":[24],"type":["chr"],"align":["left"]},{"label":["centroided"],"name":[25],"type":["lgl"],"align":["right"]},{"label":["ionMobilityDriftTime"],"name":[26],"type":["dbl"],"align":["right"]},{"label":["isolationWindowTargetMZ"],"name":[27],"type":["dbl"],"align":["right"]},{"label":["isolationWindowLowerOffset"],"name":[28],"type":["dbl"],"align":["right"]},{"label":["isolationWindowUpperOffset"],"name":[29],"type":["dbl"],"align":["right"]},{"label":["scanWindowLowerLimit"],"name":[30],"type":["dbl"],"align":["right"]},{"label":["scanWindowUpperLimit"],"name":[31],"type":["dbl"],"align":["right"]}],"data":[{"1":"8","2":"8","3":"2","4":"1","5":"49","6":"96117.32","7":"2.14157","8":"301.14122","9":"48492.03","10":"25","11":"0","12":"50.16715","13":"319.0712","14":"7","15":"301.1405","16":"1","17":"411267.5","18":"NA","19":"NA","20":"NA","21":"NA","22":"0","23":"NA","24":"controllerType=0 controllerNumber=1 scan=8","25":"TRUE","26":"NA","27":"NA","28":"1","29":"1","30":"NA","31":"NA","_rn_":"8"},{"1":"29","2":"29","3":"2","4":"1","5":"65","6":"287236.06","7":"7.81044","8":"217.10517","9":"163613.17","10":"25","11":"0","12":"50.93454","13":"218.1087","14":"28","15":"217.1069","16":"1","17":"238682.0","18":"NA","19":"NA","20":"NA","21":"NA","22":"0","23":"NA","24":"controllerType=0 controllerNumber=1 scan=29","25":"TRUE","26":"NA","27":"NA","28":"1","29":"1","30":"NA","31":"NA","_rn_":"29"},{"1":"41","2":"41","3":"2","4":"1","5":"100","6":"1545867.50","7":"11.12990","8":"113.96369","9":"626562.44","10":"25","11":"0","12":"50.89707","13":"156.1382","14":"40","15":"154.9900","16":"0","17":"89905.9","18":"NA","19":"NA","20":"NA","21":"NA","22":"0","23":"NA","24":"controllerType=0 controllerNumber=1 scan=41","25":"TRUE","26":"NA","27":"NA","28":"1","29":"1","30":"NA","31":"NA","_rn_":"41"},{"1":"81","2":"81","3":"2","4":"1","5":"104","6":"475490.09","7":"21.90190","8":"84.95964","9":"102565.70","10":"25","11":"0","12":"51.17863","13":"168.1384","14":"80","15":"167.0125","16":"1","17":"263485.1","18":"NA","19":"NA","20":"NA","21":"NA","22":"0","23":"NA","24":"controllerType=0 controllerNumber=1 scan=81","25":"TRUE","26":"NA","27":"NA","28":"1","29":"1","30":"NA","31":"NA","_rn_":"81"},{"1":"84","2":"84","3":"2","4":"1","5":"55","6":"95975.35","7":"22.66680","8":"301.14101","9":"39457.93","10":"25","11":"0","12":"51.32985","13":"322.4011","14":"83","15":"301.1404","16":"1","17":"142265.4","18":"NA","19":"NA","20":"NA","21":"NA","22":"0","23":"NA","24":"controllerType=0 controllerNumber=1 scan=84","25":"TRUE","26":"NA","27":"NA","28":"1","29":"1","30":"NA","31":"NA","_rn_":"84"},{"1":"100","2":"100","3":"2","4":"1","5":"46","6":"99555.92","7":"27.00660","8":"299.11010","9":"58150.55","10":"25","11":"0","12":"50.60672","13":"305.5677","14":"99","15":"299.1095","16":"1","17":"280681.7","18":"NA","19":"NA","20":"NA","21":"NA","22":"0","23":"NA","24":"controllerType=0 controllerNumber=1 scan=100","25":"TRUE","26":"NA","27":"NA","28":"1","29":"1","30":"NA","31":"NA","_rn_":"100"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>The next useful function retrieves the actual mass spectra. It will load every scan in the file as a two-column separate matrix. For each matrix the first column represents m/z and the second column is intensity.</p>
<p>Let’s look at the first five lines of the twenty-second scan/mass spectrum.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># note: mzR::peaks() and mzR::spectra() are interchangeable</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>single_spectrum <span class="ot">&lt;-</span> mzR<span class="sc">::</span><span class="fu">peaks</span>(msfile_handle, <span class="at">scans=</span><span class="dv">4</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(single_spectrum, <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         [,1]      [,2]
[1,] 150.0265 32913.336
[2,] 151.0238  2110.815
[3,] 151.0272  3636.793
[4,] 152.0564  4872.385
[5,] 153.0907  2387.040</code></pre>
</div>
</div>
<p>We can load just the filtered MS<sup>2</sup></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>positive_ms2_header <span class="ot">&lt;-</span> summary_data[summary_data<span class="sc">$</span>polarity <span class="sc">==</span> <span class="dv">1</span>, ][summary_data<span class="sc">$</span>msLevel <span class="sc">==</span> <span class="dv">2</span>, ]</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>positive_ms2_scan_index <span class="ot">&lt;-</span> positive_ms2_header<span class="sc">$</span>seqNum</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>positive_ms2_spectra <span class="ot">&lt;-</span> mzR<span class="sc">::</span><span class="fu">peaks</span>(msfile_handle, positive_ms2_scan_index)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Check that we only grabbed the positive, MS<sup>2</sup> spectra</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(positive_ms2_spectra) <span class="sc">==</span> <span class="fu">nrow</span>(positive_ms2_header)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>selected_scan <span class="ot">&lt;-</span> <span class="dv">2243</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>spectrum_to_plot <span class="ot">&lt;-</span> mzR<span class="sc">::</span><span class="fu">peaks</span>(msfile_handle, selected_scan)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>spectrum_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(spectrum_to_plot)</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(spectrum_df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"m/z"</span>, <span class="st">"intensity"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s write a couple helper functions first.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Fancy_scientific creates scientific notation text</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>fancy_scientific <span class="ot">&lt;-</span> <span class="cf">function</span>(l) {</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># "fancy_scientific" was copied from comment from Brian Diggs posted at https://groups.google.com/forum/#!topic/ggplot2/a_xhMoQyxZ4</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># turn in to character string in scientific notation</span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>  l <span class="ot">&lt;-</span> <span class="fu">format</span>(l, <span class="at">scientific =</span> <span class="cn">TRUE</span>)</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># quote the part before the exponent to keep all the digits</span></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>  l <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"^(.*)e"</span>, <span class="st">"'</span><span class="sc">\\</span><span class="st">1'e"</span>, l)</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># turn the 'e+' into plotmath format</span></span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>  l <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"e"</span>, <span class="st">"%*%10^"</span>, l)</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># return this as an expression</span></span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">parse</span>(<span class="at">text=</span>l)</span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a><span class="co"># gets top N intensity peaks</span></span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a>topn <span class="ot">&lt;-</span> <span class="cf">function</span>(x, <span class="at">n=</span><span class="dv">10</span>) { x[<span class="fu">order</span>(x[, <span class="dv">2</span>], <span class="at">decreasing =</span> <span class="fu">ifelse</span>(<span class="fu">all</span>(x[, <span class="dv">2</span>] <span class="sc">&lt;</span> <span class="dv">0</span>), F, T))[<span class="dv">1</span><span class="sc">:</span>n], ] }</span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-17"><a href="#cb34-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-18"><a href="#cb34-18" aria-hidden="true" tabindex="-1"></a>plot_peaks <span class="ot">&lt;-</span> <span class="cf">function</span>(df1, plot_title){</span>
<span id="cb34-19"><a href="#cb34-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb34-20"><a href="#cb34-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">data =</span> df1, <span class="fu">aes</span>(<span class="at">x =</span> <span class="st">`</span><span class="at">m/z</span><span class="st">`</span>, <span class="at">y =</span> <span class="st">`</span><span class="at">intensity</span><span class="st">`</span>), <span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">width =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb34-21"><a href="#cb34-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">data=</span><span class="fu">topn</span>(df1), <span class="fu">aes</span>(<span class="at">x =</span> <span class="st">`</span><span class="at">m/z</span><span class="st">`</span>, <span class="at">y =</span> <span class="st">`</span><span class="at">intensity</span><span class="st">`</span> <span class="sc">*</span> <span class="fl">1.05</span>, <span class="at">label =</span> <span class="fu">as.numeric</span>(<span class="fu">sprintf</span>(<span class="st">'%.4f'</span>, <span class="st">`</span><span class="at">m/z</span><span class="st">`</span>))), <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb34-22"><a href="#cb34-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">name=</span><span class="st">"Intensity"</span>, <span class="at">labels =</span> fancy_scientific) <span class="sc">+</span></span>
<span id="cb34-23"><a href="#cb34-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="at">label =</span> <span class="fu">expression</span>(<span class="fu">italic</span>(<span class="st">"m/z"</span>))) <span class="sc">+</span></span>
<span id="cb34-24"><a href="#cb34-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(plot_title) <span class="sc">+</span></span>
<span id="cb34-25"><a href="#cb34-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span>
<span id="cb34-26"><a href="#cb34-26" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb34-27"><a href="#cb34-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-28"><a href="#cb34-28" aria-hidden="true" tabindex="-1"></a>plot_mirror <span class="ot">&lt;-</span> <span class="cf">function</span>(df1, df2, plot_title, df1_name, df2_name){</span>
<span id="cb34-29"><a href="#cb34-29" aria-hidden="true" tabindex="-1"></a>  df1<span class="sc">$</span><span class="st">`</span><span class="at">intensity</span><span class="st">`</span> <span class="ot">&lt;-</span> df1<span class="sc">$</span><span class="st">`</span><span class="at">intensity</span><span class="st">`</span> <span class="sc">/</span> <span class="fu">max</span>(df1<span class="sc">$</span><span class="st">`</span><span class="at">intensity</span><span class="st">`</span>) <span class="sc">*</span> <span class="dv">100</span> </span>
<span id="cb34-30"><a href="#cb34-30" aria-hidden="true" tabindex="-1"></a>  df2<span class="sc">$</span><span class="st">`</span><span class="at">intensity</span><span class="st">`</span> <span class="ot">&lt;-</span> df2<span class="sc">$</span><span class="st">`</span><span class="at">intensity</span><span class="st">`</span> <span class="sc">/</span> <span class="fu">max</span>(df2<span class="sc">$</span><span class="st">`</span><span class="at">intensity</span><span class="st">`</span>) <span class="sc">*</span> <span class="dv">100</span> <span class="sc">*</span> <span class="sc">-</span><span class="dv">1</span></span>
<span id="cb34-31"><a href="#cb34-31" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb34-32"><a href="#cb34-32" aria-hidden="true" tabindex="-1"></a>  df3 <span class="ot">&lt;-</span> <span class="fu">rbind.data.frame</span>(</span>
<span id="cb34-33"><a href="#cb34-33" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cbind.data.frame</span>(df1, <span class="at">spectrum=</span>df1_name),</span>
<span id="cb34-34"><a href="#cb34-34" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cbind.data.frame</span>(df2, <span class="at">spectrum=</span>df2_name)</span>
<span id="cb34-35"><a href="#cb34-35" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb34-36"><a href="#cb34-36" aria-hidden="true" tabindex="-1"></a>  <span class="co"># ensure the input spectrum (positive spectrum) appears as the top label in the legend</span></span>
<span id="cb34-37"><a href="#cb34-37" aria-hidden="true" tabindex="-1"></a>  df3<span class="sc">$</span>spectrum <span class="ot">&lt;-</span> <span class="fu">factor</span>(df3<span class="sc">$</span>spectrum, <span class="at">levels=</span><span class="fu">c</span>(df1_name, df2_name))</span>
<span id="cb34-38"><a href="#cb34-38" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb34-39"><a href="#cb34-39" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb34-40"><a href="#cb34-40" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>(<span class="at">data =</span> df3, <span class="fu">aes</span>(<span class="at">x =</span> <span class="st">`</span><span class="at">m/z</span><span class="st">`</span>, <span class="at">y =</span> <span class="st">`</span><span class="at">intensity</span><span class="st">`</span>, <span class="at">fill=</span><span class="st">`</span><span class="at">spectrum</span><span class="st">`</span>), <span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">width =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb34-41"><a href="#cb34-41" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="st">"legend"</span>, <span class="at">values =</span> <span class="fu">c</span>( <span class="st">"blue"</span>,  <span class="st">"red"</span>)) <span class="sc">+</span></span>
<span id="cb34-42"><a href="#cb34-42" aria-hidden="true" tabindex="-1"></a>    <span class="co"># write the m/z labels to 4 decimal places</span></span>
<span id="cb34-43"><a href="#cb34-43" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_text</span>(<span class="at">data=</span><span class="fu">topn</span>(df1), <span class="fu">aes</span>(<span class="at">x =</span> <span class="st">`</span><span class="at">m/z</span><span class="st">`</span>, <span class="at">y =</span> <span class="st">`</span><span class="at">intensity</span><span class="st">`</span> <span class="sc">*</span> <span class="fl">1.05</span>, <span class="at">label =</span> <span class="fu">as.numeric</span>(<span class="fu">sprintf</span>(<span class="st">'%.4f'</span>, <span class="st">`</span><span class="at">m/z</span><span class="st">`</span>))), <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb34-44"><a href="#cb34-44" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_text</span>(<span class="at">data=</span><span class="fu">topn</span>(df2), <span class="fu">aes</span>(<span class="at">x =</span> <span class="st">`</span><span class="at">m/z</span><span class="st">`</span>, <span class="at">y =</span> <span class="st">`</span><span class="at">intensity</span><span class="st">`</span> <span class="sc">*</span> <span class="fl">1.05</span>, <span class="at">label =</span> <span class="fu">as.numeric</span>(<span class="fu">sprintf</span>(<span class="st">'%.4f'</span>, <span class="st">`</span><span class="at">m/z</span><span class="st">`</span>))), <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb34-45"><a href="#cb34-45" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_y_continuous</span>(<span class="at">name=</span><span class="st">"Intensity"</span>, <span class="at">labels =</span> fancy_scientific) <span class="sc">+</span></span>
<span id="cb34-46"><a href="#cb34-46" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlab</span>(<span class="at">label =</span> <span class="fu">expression</span>(<span class="fu">italic</span>(<span class="st">"m/z"</span>))) <span class="sc">+</span></span>
<span id="cb34-47"><a href="#cb34-47" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(plot_title) <span class="sc">+</span></span>
<span id="cb34-48"><a href="#cb34-48" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span>
<span id="cb34-49"><a href="#cb34-49" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>input_file_name <span class="ot">&lt;-</span> <span class="fu">basename</span>(mzR<span class="sc">::</span><span class="fu">fileName</span>(msfile_handle))</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>plot_title <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"File: "</span>, input_file_name, <span class="st">"; scan: "</span>, selected_scan)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_peaks</span>(spectrum_df, plot_title)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: `position_stack()` requires non-overlapping x intervals</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>gnps_spectrum_df <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">"/home/chase/Downloads/CCMSLIB00000072054.mgf"</span>, <span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">skip =</span> <span class="dv">3</span>, <span class="at">header=</span><span class="cn">FALSE</span>, <span class="at">nrows =</span> <span class="dv">500</span>)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>gnps_spectrum_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(gnps_spectrum_df)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(gnps_spectrum_df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"m/z"</span>, <span class="st">"intensity"</span>)</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>plot_title <span class="ot">&lt;-</span> <span class="st">"GNPS Library: CCMSLIB00000072054"</span></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_peaks</span>(gnps_spectrum_df, plot_title)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: `position_stack()` requires non-overlapping x intervals</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_mirror</span>(</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">df1 =</span> spectrum_df,</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">df2 =</span> gnps_spectrum_df,</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">df1_name =</span> <span class="st">"input"</span>,</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">df2_name =</span> <span class="st">"CCMSLIB00000072054"</span>,</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot_title =</span> <span class="st">"Comparison to GNPS Reference"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: `position_stack()` requires non-overlapping x intervals
`position_stack()` requires non-overlapping x intervals</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      return note.innerHTML;
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<script src="https://giscus.app/client.js" data-repo="quarto-dev/quarto-docs" data-repo-id="" data-category="General" data-category-id="" data-mapping="title" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="light" data-lang="en" crossorigin="anonymous" async="">
</script>
<input type="hidden" id="giscus-base-theme" value="light">
<input type="hidden" id="giscus-alt-theme" value="dark">
</div> <!-- /content -->




</body></html>